### PR TITLE
allow using Iterable in Series and DataFrame constructor

### LINKED ISSUE #2193: Cannot produce DataFrame from dict-yielding iterator.
Hi there,

I have a codebase that (so far) consumes data series in the form of dict-yielding iterators, similar in form to how DataFrame's constructor accepts a list of such. Unfortunately it seems from current implementation of that constructor, there is no straightforward way to consume from an iterator straight into a DataFrame.

For example, I have a row factory for sqlite3 written in C that produces dicts. If pandas would accept an iterator, it should be possible to consume a series directly from speedy storage code into speedy data structure code. :)

Is there an inherent limitation preventing this? Otherwise with a few pointers I would be willing to try contribute the patch myself.

Thanks,

David

### PR DESCRIPTION
**Pull request by holymonson**
- [X] closes #2193
- [x] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry
---

@TomAugspurger Hope you can review this.

BTW, perhpas you may want to change `Iterable` in `is_list_like()` to `Sequence`.

```python
gen = (i for i in range(10))
pandas.core.dtypes.inference.is_list_like(gen)
# True
len(gen)
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
# TypeError: object of type 'generator' has no len()
gen[0]
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
# TypeError: 'generator' object is not subscriptable
```

### CODE REVIEW

**Comment by holymonson** [comment]
This test failed. But a single-column `DataFrame` seems reasonable to me.
https://github.com/pandas-dev/pandas/blob/27ebb3e1e40513ad5f8919a5bbc7298e2e070a39/pandas/tests/frame/test_constructors.py#L868-L872
```python
pd.DataFrame((1, 2, 3))
#   0
#0  1
#1  2
#2  3
```

File: pandas/tests/series/test_constructors.py [hunk 1]
```diff
@@ -156,10 +156,20 @@ def test_constructor_series(self):
...
-    def test_constructor_iterator(self):
+    def test_constructor_iterable(self):
+        class Iter():
+            def __iter__(self):
+                for i in range(10):
+                    yield i
 
         expected = Series(list(range(10)), dtype='int64')
-        result = Series(range(10), dtype='int64')
+        result = Series(Iter(), dtype='int64')
```
Comment (mroeschke) [inline pandas/tests/series/test_constructors.py]: cant this simply be iter(range(10))?

Comment (holymonson) [reply pandas/tests/series/test_constructors.py]: `range` is list-like, subclass of `Sequence`, which doesn't need  consuming by `data = list(data)`. I'd like to add back `range(10)` test and keep both.

**Comment by holymonson** [comment]
Allow `pd.DataFrame((1, 2, 3))` because `pd.DataFrame([1, 2, 3])` is valid. They should be the same.

File: pandas/tests/frame/test_constructors.py [hunk 2]
```diff
@@ -953,6 +947,16 @@ def __len__(self, n):
...
+    def test_constructor_iterable(self):
+        class Iter():
+            def __iter__(self):
+                for i in range(10):
+                    yield [1, 2, 3]
+
+        expected = DataFrame([[1, 2, 3]] * 10)
+        result = DataFrame(Iter())
+        tm.assert_frame_equal(result, expected)
+
```
Comment (gfyoung) [inline pandas/tests/frame/test_constructors.py]: Reference the issue number in a comment below the functional definition.  Applies to all newly-added tests.

**Comment by TomAugspurger** [comment]
> Allow pd.DataFrame((1, 2, 3)) because pd.DataFrame([1, 2, 3]) is valid. They should be the same.

Context: That goes back a long way to DataMatrix: https://github.com/pandas-dev/pandas/commit/c34ac749382c2046548eb2470eb847d66ecc6d23#diff-225c6e44f997b9b698ab52a48b223018R39

+1 for allowing it.

**Review by TomAugspurger — APPROVED** [review]
Merging later today, unless someone else beats me to it.

**Comment by jreback** [comment]
will have a look

don’t merge

File: pandas/core/frame.py [hunk 3]
```diff
@@ -390,8 +390,9 @@ def __init__(self, data=None, index=None, columns=None, dtype=None,
             else:
                 mgr = self._init_ndarray(data, index, columns, dtype=dtype,
                                          copy=copy)
-        elif isinstance(data, (list, types.GeneratorType)):
-            if isinstance(data, types.GeneratorType):
+        elif (isinstance(data, collections.Iterable)
```
Comment (jreback) [inline pandas/core/frame.py]: can you add a comment here (1 line above) which defines what this clause is checking

File: pandas/core/series.py [hunk 4]
```diff
@@ -238,12 +238,12 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
 
             elif is_extension_array_dtype(data):
                 pass
-            elif (isinstance(data, types.GeneratorType) or
-                  (compat.PY3 and isinstance(data, map))):
-                data = list(data)
             elif isinstance(data, (set, frozenset)):
                 raise TypeError("{0!r} type is unordered"
                                 "".format(data.__class__.__name__))
+            elif (isinstance(data, collections.Iterable)
```
Comment (jreback) [inline pandas/core/series.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew note, other enhancements in 0.24

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @holymonson !

**Pull request merged**